dtm0/dtm0-dld.c

```diff
@@ -0,0 +1,1549 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+
+/**
+   @page DLD Motr DLD Template
+
+   - @ref DLD-ovw (todo)
+   - @ref DLD-def (done)
+   - @ref DLD-req
+   - @ref DLD-depends
+   - @ref DLD-highlights
+   - @subpage DLD-fspec "Functional Specification" <!-- Note @subpage -->
+   - @ref DLD-lspec
+      - @ref DLD-lspec-comps
+      - @ref DLD-lspec-sub
+      - @ref DLD-lspec-state
+      - @ref DLD-lspec-thread
+      - @ref DLD-lspec-numa
+   - @ref DLD-conformance
+   - @ref DLD-ut
+   - @ref DLD-st
+   - @ref DLD-O
+   - @ref DLD-ref
+   - @ref DLD-impl-plan
+
+   <hr>
+   @section DLD-ovw Overview
+   <i>All specifications must start with an Overview section that
+   briefly describes the document and provides any additional
+   instructions or hints on how to best read the specification.</i>
+
+   TODO: write overview; add to the index in dld-index.c.
+
+   <hr>
+   @section DLD-def Definitions
+   <i>Mandatory.
+   The DLD shall provide definitions of the terms and concepts
+   introduced by the design, as well as the relevant terms used by the
+   specification but described elsewhere.  References to the
+   M0 Glossary and the component's HLD are permitted and encouraged.
+   Agreed upon terminology should be incorporated in the glossary.</i>
+
+   Previously defined terms:
+   - <b>Storage device</b> A configuration object that corresponds to a physical
+   device where Motr keeps data. In this document, this term is used as synonym
+   for "persistent participant": Motr stores its persistent data on persistent
+   participants.
+   - <b>Unit</b> Unit of data. In this document, this term is used
+   primary to describe DIX records (DIX/CAS) and data units (IO/IOS).
+   New terms:
+   - <b>Participant</b> A member of a distributed transaction. Participants
+   comprise originators and storage devices.
+   - <b>Originator</b> The initiator of a distributed transaction. Originators
+   has no persistent storage.
+   - <b>PERSISTENT message, Pmsg, Pmgs (plural)</b> A message that indicates
+   that a certain information (transaction, log record) became persistent
+   on a certain storage device.
+   <b>Log record has/is All-P</b>: P messages were received about all non-FAILED
+   storage devices that are participants of this log record's dtx.
```

Comment:
note, "non-FAILED". It is a reference to M0_NC_FAILED. In this context, "non-FAILED" means the devices that did not experience permanent failure (M0_NC_FAILED).

If a device enters M0_NC_FAILED then we won't be able to REDO any transaction where the sdev is a participant with missing P message. Because of that we do not consider permanently failed devices when we "calculate" All-P.